File: app/src/main/java/com/example/ui/alarm/AlarmStateHolder.java
package com.example.ui.alarm;

import com.example.data.model.Alarm;
import kotlinx.coroutines.flow.MutableStateFlow;
import kotlinx.coroutines.flow.StateFlow;
import kotlinx.coroutines.flow.StateFlowKt;

public class AlarmStateHolder {
    private static final MutableStateFlow<Alarm> _ringingAlarm = StateFlowKt.MutableStateFlow(null);
    public static final StateFlow<Alarm> ringingAlarm = _ringingAlarm;

    private static final MutableStateFlow<FocusState> _focusModeState = StateFlowKt.MutableStateFlow(null);
    public static final StateFlow<FocusState> focusModeState = _focusModeState;

    public static void startRinging(Alarm alarm) {
        _ringingAlarm.setValue(alarm);
    }

    public static void stopRinging() {
        _ringingAlarm.setValue(null);
    }

    public static void startFocusMode(
        int alarmId,
        int durationMinutes,
        String taskType,
        String taskDetails,
        String referencePhotoPath
    ) {
        long totalMillis = durationMinutes * 60 * 1000L;
        long endTime = System.currentTimeMillis() + totalMillis;
        _focusModeState.setValue(new FocusState(
            alarmId,
            endTime,
            Math.max(totalMillis, 1000L),
            taskType,
            taskDetails,
            referencePhotoPath
        ));
    }

    public static void clearFocusMode() {
        _focusModeState.setValue(null);
    }
}

File: app/src/main/java/com/example/ui/alarm/AlarmStateHolder.kt
package com.example.ui.alarm

import com.example.data.model.Alarm
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

object AlarmStateHolder {
    private val _ringingAlarm = MutableStateFlow<Alarm?>(null)
    val ringingAlarm: StateFlow<Alarm?> = _ringingAlarm.asStateFlow()

    private val _focusModeState = MutableStateFlow<FocusState?>(null)
    val focusModeState: StateFlow<FocusState?> = _focusModeState.asStateFlow()

    fun startRinging(alarm: Alarm) {
        _ringingAlarm.value = alarm
    }

    fun stopRinging() {
        _ringingAlarm.value = null
    }

    fun startFocusMode(
        alarmId: Int,
        durationMinutes: Int,
        taskType: String,
        taskDetails: String,
        referencePhotoPath: String? = null
    ) {
        val totalMillis = durationMinutes * 60 * 1000L
        val endTime = System.currentTimeMillis() + totalMillis
        _focusModeState.value = FocusState(
            alarmId = alarmId,
            endTimeMillis = endTime,
            totalDurationMillis = totalMillis.coerceAtLeast(1000L),
            taskType = taskType,
            taskDetails = taskDetails,
            referencePhotoPath = referencePhotoPath
        )
    }

    fun clearFocusMode() {
        _focusModeState.value = null
    }
}

data class FocusState(
    val alarmId: Int,
    val endTimeMillis: Long,
    val totalDurationMillis: Long,
    val taskType: String,
    val taskDetails: String,
    val referencePhotoPath: String? = null
)
